packages/flutter/lib/src/material/tab_controller.dart

```diff
@@ -0,0 +1,201 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'constants.dart';
+
+/// Coordinates tab selection between a [TabBar] and a [TabBarView].
+///
+/// The [index] property is the index of the selected tab and the [animation]
+/// represents the current scroll positions of the tab bar and the tar bar view.
+/// The selected tab's index can be changed with [animateTo].
+///
+/// See also:
+///
+/// * [DefaultTabController], which simplifies sharing a TabController with
+/// its [TabBar] and a [TabBarView] descendants.
+class TabController extends ChangeNotifier {
+  /// Creates an object that manages the state required by [TabBar] and a [TabBarView].
+  TabController({ int initialIndex: 0, @required this.length, @required TickerProvider vsync })
+    : _index = initialIndex,
+      _previousIndex = initialIndex,
+      _animationController = new AnimationController(
+        value: initialIndex.toDouble(),
+        upperBound: (length - 1).toDouble(),
+        vsync: vsync
+   ) {
+    assert(length != null && length > 1);
+    assert(initialIndex != null && initialIndex >= 0 && initialIndex < length);
+  }
+
+  /// An animation whose value represents the current position of the [TabBar]'s
+  /// selected tab indicator as well as the scrollOffsets of the [TabBar]
+  /// and [TabBarView].
+  ///
+  /// The animation's value ranges from 0.0 to [length] - 1.0. After the
+  /// selected tab is changed, the animation's value equals [index]. The
+  /// animation's value can be [offset] by +/- 1.0 to reflect [TabBarView]
+  /// drag scrolling.
+  final AnimationController _animationController;
+  Animation<double> get animation => _animationController.view;
+
+  /// The total number of tabs. Must be greater than one.
+  final int length;
+
+  void _changeIndex(int value, { Duration duration, Curve curve }) {
+    assert(value != null);
+    assert(value >= 0 && value < length);
+    assert(duration == null ? curve == null : true);
+    if (value == _index)
+      return;
+    _indexIsChanging = true;
+    _previousIndex = index;
+    _index = value;
+    if (duration != null) {
+      _animationController
+        ..animateTo(_index.toDouble(), duration: duration, curve: curve).then((_) {
+          _indexIsChanging = false;
```

Comment:
Excellent point!  I made a similar mistake with the _warpUnderway bool as you pointed out below.